docs/API.md

```diff
@@ -0,0 +1,267 @@
+
+#### custom
+
+The `custom` property can be used to attach custom data to the `App` which will be exposed at `App.custom`.  This data can be used in custom initialization handlers, or elsewhere in the application as necessary.  Note, if you're using this to provide mutable data to the application, _strongly_ consider using React props, context, or Redux instead.
+
+### `App.config`
+
+The environment configuration.  Contains the following keys:
+
+- ACCESS_TOKEN_COOKIE_NAME
+- BASE_URL
+- CREDENTIALS_BASE_URL
+- CSRF_COOKIE_NAME
+- CSRF_TOKEN_API_PATH
+- ECOMMERCE_BASE_URL
+- ENVIRONMENT
+- LANGUAGE_PREFERENCE_COOKIE_NAME
+- LMS_BASE_URL
+- LOGIN_URL
+- LOGOUT_URL
+- MARKETING_SITE_BASE_URL
+- ORDER_HISTORY_URL
+- REFRESH_ACCESS_TOKEN_ENDPOINT
+- SECURE_COOKIES
+- SEGMENT_KEY
+- SITE_NAME
+- USER_INFO_COOKIE_NAME
+
+If additional, dynamic config is desired, it would be reasonable to add those keys into `App.config`.
+
+### `App.apiClient`
+
+A reference to the @edx/frontend-auth authenticated API Client.
+
+### `App.subscribe(type, callback)`
+
+A method allowing consumers of `App` to subscribe to lifecycle events.  `type` is an event type, as documented in "Initialization Lifecycle Phases".  There are constants for all the event types:
+
+```
+import {
+  APP_BEFORE_INIT, APP_CONFIGURED, APP_AUTHENTICATED, APP_I18N_CONFIGURED, APP_LOGGING_CONFIGURED, APP_ANALYTICS_CONFIGURED, APP_BEFORE_READY, APP_READY, APP_ERROR
+} from `@edx/frontend-base`
+```
+
+### `App.requireConfig(keys, requester)`
+
+A method allowing application code to indicate that particular `App.config` keys are required for them to function.  The method returns the required config values.
+
+```
+const config = App.requireConfig(['LMS_BASE_URL', 'LOGIN_URL'], 'The name of my feature/component');
+```
+
+**NOTE**: If you use App.requireConfig to require config that hasn't been loaded yet (i.e., from a custom `configuration` handler), it will throw an error.  If you use a custom `configuration` handler, you can defer requiring your config by subscribing to the `APP_CONFIGURED` event:
+
+```
+let config = null;
+App.subscribe(APP_CONFIGURED, () => {
+  config = App.requireConfig(['DYNAMICALLY_CONFIGURED_URL'], 'Consumer of custom config');
+
+  // config is known to be set here.
+});
+
+export function usesConfig() {
+  /* If this function is not used before APP_CONFIGURED, then this is safe usage.
+   * If you expect your method may be used before APP_CONFIGURED, then you have a
+   * more complex setup.  That sort of thing should be very uncommon.
+   */
+  console.log(config.DYNAMICALLY_CONFIGURED_URL);
+}
+```
+
+Environment variable-based config is immediately available on file load, so generally waiting for APP_CONFIGURED is not necessary.
+
+### `App.queryParams`
+
+A method which converts the current query string into an object of key-value pairs and returns it.  It is memoized based on the current query string for efficiency.
```

Comment:
Should we note that we expect this to move to a separate library?  Can we move it now?  Or, do we want helpers to be added here, since you are setting a precedence.

Reply:
I wasn't intending to move this anywhere, but maybe you're right... I don't want to set a precedent of adding utility methods to `App` - there are better ways to do that.  

This was here because it's a sorta "top level" thing that the app is easing access to; similar to what `history` does, though that's a third party module.  I think for now I'll just leave it, but want to be mindful of other things getting added here...